A logarithmic algorithm to find the next free variable

gasche opened the pull request:
Several places in the compiler have logic to "pick a fresh name" by starting from a prefix, say foo, and trying foo/0, foo/1, foo/2... until they find one that is "available" in some sense.

This can easily be implemented through a linear search, but this gives a quadratic behavior when we repeatedly generate fresh names from the same prefix -- which can happen in practice in realistic example of source code repeating the same variable or constructor name many times.

We introduce a function `Misc.find_first_mono` that finds the "first available number" in logarithmic time, assuming that the "unavailable numbers" form a strict prefix of all natural numbers. (That is: some numbers remain available, and all numbers after the first available numbers are available.)

When called repeatedly on the same prefix, this gives a behavior in O(n log n) rather than O(n^2).

gasche committed:
Misc.find_first_mono

Several places in the compiler have logic to "pick a fresh name"
by starting from a prefix, say foo, and trying foo/0, foo/1,
foo/2... until they find one that is "available" in some sense.

This can easily be implemented through a linear search, but this gives
a quadratic behavior when we repeatedly generate fresh names from the
same prefix -- which can happen in practice in realistic example of
source code repeating the same variable or constructor name many
times.

We introduce a function that finds the "first available number" in
logarithmic time, assuming that the "unavailable numbers" form
a strict prefix of all natural numbers. (That is: some numbers remain
available, and all numbers after the first available numbers are
available.)

When called repeatedly on the same prefix, this gives a behavior in
O(n log n) rather than O(n^2).

Note: if called on availability predicates that are not
monotonous (they become false again after being true), the function
will not necessarily provide the smallest available number, but it
should still provide an available number.

gasche committed:
use find_first_mono in Printtyp

gasche committed:
use find_first_mono in Untypeast

gasche committed:
Changes

gasche committed:
.depend

gasche commented:
More information, because why not:

- I checked that the function is not slower in practice than linear search when the first available number is small. For n=2 (when the first available number is 2), it performs 4 calls to the predicate instead of 3 for linear search. For n=5, it performs 7 calls instead of 6 for linear search. For all other inputs it does as well or better than linear search, and it does strictly better starting at n=7.

- If for some reason the availability predicate is not monotonous, that is, there are some "unavailable" numbers in the middle of the range of available numbers, the function still behaves reasonably as long as the predicate remains constantly true for large enough values. (It may not return the smallest available number, but it returns an available number instead of looping or crashing.)

smuenzel commented on utils/misc.ml:
```diff
@@ -395,6 +395,39 @@ module Int_literal_converter = struct
+(* [find_first_mono p] assumes that there exists a natural number
+   N such that [p] is false on [0; N[ and true on [N; max_int], and
+   returns this N. (See misc.mli for the detailed specification.) *)
+let find_first_mono p =
```
What about
```
let find_first_mono =
	let rec find ....
  ....
  in
  fun p ->
```
to ensure syntactically that we don't capture any variables in `find`. It's probably not critical for one parameter, but I find it very useful.

gasche replied:
done

smuenzel commented on utils/misc.ml:
```diff
@@ -395,6 +395,39 @@ module Int_literal_converter = struct
+   N such that [p] is false on [0; N[ and true on [N; max_int], and
+   returns this N. (See misc.mli for the detailed specification.) *)
+let find_first_mono p =
+  let rec find p low jump high =
```
Since we have three integer arguments, can we make them labeled so that we don't accidentally swap them?

smuenzel commented on utils/misc.ml:
```diff
@@ -395,6 +395,39 @@ module Int_literal_converter = struct
+    if low + 1 = high then high
+    (* ensure that [low + jump] is in ]low; high[ *)
+    else if jump < 1 then find p low 1 high
+    else if jump >= high - low then find p low ((high - low) / 2) high
```
maybe some aliases will help here?
e.g.
`let interval_size = high - low and candidate_point = low + jump in`

gasche replied:
meh ; I don't see a nice way to do this without introducing unnecessary computations in some branches (which is fine performance-wise but still slightly frustrating), and I don't think that the readability gains are that large in practice, so I did not apply this proposal

smuenzel commented on utils/misc.ml:
```diff
@@ -395,6 +395,39 @@ module Int_literal_converter = struct
+    (* ensure that [low + jump] is in ]low; high[ *)
+    else if jump < 1 then find p low 1 high
+    else if jump >= high - low then find p low ((high - low) / 2) high
+    else if not (p (low + jump)) then
```
Why have the negated condition first?

gasche replied:
I prefer to have the simpler case first, always. This said, now that I removed the "sequential search half the time" component, the "too high" case is in fact just as simple or simpler, so I did end up switching the two cases.

smuenzel commented on utils/misc.ml:
```diff
@@ -395,6 +395,39 @@ module Int_literal_converter = struct
+       [low, jump, high] are non-negative with [low < high],
+       [p low = false],
+       [p high = true]. *)
+    if low + 1 = high then high
```
There seem to be two parts of the if tree: 
1. Validate interval/jump size
2. Evaluate the candidate point
I'm not sure exactly how, but maybe separating the two would help with understanding?

gasche replied:
I tried at first to have logic to validate the jump size in each case (too high / too low), but the code ended up being less clear, with more invariants to state to reason about correctness. I would rather keep the current approach.

smuenzel commented on utils/misc.mli:
```diff
@@ -307,6 +307,25 @@ module Int_literal_converter : sig
+val find_first_mono : (int -> bool) -> int
+  (**[find_first_mono p] takes an integer predicate [p : int -> bool]
+     that we assume:
+     1. is monotonous on natural numbers:
```
monotonic, not monotonous, I think

gasche replied:
thanks!

smuenzel commented on typing/printtyp.ml:
```diff
@@ -936,16 +936,16 @@ end = struct
-            done;
-            !current_name
+            let available name =
+              List.exists
```
I guess we're still linear in the length of this list, but it doesn't matter in practice?

smuenzel replied:
I guess the real question is, how large does the `names` list get, and if it gets large, maybe we should change from a list to a different data structure.

gasche replied:
I consider this as orthogonal to the use of `find_first_mono` in this function, although you are right that we probably don't get a good worst-case complexity in this case. I'm okay with that -- I propose to introduce `find_first_mono` and use it everywhere applicable, even if it does not solve all complexity issues.

smuenzel commented:
The build shows some diff/bug:
```
> type foo = Foo : 'b4611686018427387903 * 'b4611686018427387903 -> foo
```

smuenzel commented on typing/printtyp.ml:
```diff
@@ -936,16 +936,16 @@ end = struct
-              i := !i + 1;
-            done;
-            !current_name
+            let available name =
```
I think this is `unavailable`. Maybe add a `not`?

gasche replied:
I guess that this is the bug :-)

gasche commented:
Thanks @smuenzel for the nice review. I updated the PR to follow (most of) your recommendations. I also took this opportunity to check that the "sequential half the time" strategy does in fact improve things, and in fact it does not, so I simplified the implementation accordingly.

smuenzel reviewed (approved):
Looks great, and seems like the checks are happy now too!

gasche commented:
Thanks! I would be inclined to "approve on @smuenzel's behalf", but I am hesitant to do this for my own PR. I'll leave the decision up to other maintainers.

lthls reviewed (approved):
I've read enough to convince me this PR makes sense, and trust @smuenzel for the fine details.

gasche commented:
Thanks! Merging.
